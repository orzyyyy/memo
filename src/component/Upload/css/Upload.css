.upload .upload-item {
  border: 1px dashed #d9d9d9;
  width: 26vw;
  height: 26vw;
  border-radius: 4px;
  background-color: #fafafa;
  text-align: center;
  cursor: pointer;
  transition: border-color 0.3s ease;
  vertical-align: top;
  display: table;
  padding: 5px;
  margin: 1vw;
}

.upload .upload-item-error {
  border: 1px dashed red;
}
