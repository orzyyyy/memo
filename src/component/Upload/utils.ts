const imageTypes = ['image', 'webp', 'png', 'svg', 'gif', 'jpg', 'jpeg', 'bmp'];

const extname = (url?: string) => {
  if (!url) {
    return '';
  }
  const temp = url.split('/');
  const filename = temp[temp.length - 1];
  const filenameWithoutSuffix = filename.split(/#|\?/)[0];
  return (/\.[^./\\]*$/.exec(filenameWithoutSuffix) || [''])[0];
};

export const isImageUrl = (file: any) => {
  if (!file) {
    return false;
  }
  if (file.type && imageTypes.includes(file.type)) {
    return true;
  }
  const url = file.url;
  const extension = extname(url);
  if (/^data:image\//.test(url) || /(webp|svg|png|gif|jpg|jpeg|bmp)$/i.test(extension)) {
    return true;
  } else if (/^data:/.test(url)) {
    return false;
  } else if (extension) {
    return false;
  }
  return true;
};
