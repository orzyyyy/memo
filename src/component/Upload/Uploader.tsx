import React, { useRef } from 'react';
import './css/Uploader.css';
import { isImageUrl } from './utils';

export interface UploadFile {
  id: string;
  url: string;
  name: string;
  error?: boolean;
}
export interface UploaderProps {
  onChange?: (file: UploadFile) => void;
  accept?: string;
  multiple?: boolean;
  renderPlusItem?: () => void;
}

const Uploader = ({ renderPlusItem, multiple, onChange, ...rest }: UploaderProps) => {
  let uploadInput: React.RefObject<HTMLInputElement> = useRef(null);

  const triggerInputUpload = () => {
    const element = uploadInput.current;
    if (!element) {
      return;
    }
    element.click();
    element.value = '';
  };

  const onFileChange = (e: any, callback?: Function) => {
    const file = e.target.files.length ? e.target.files[0] : null;
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onloadend = () => {
      const result = { url: reader.result };
      file.url = null;
      if (isImageUrl(result)) {
        file.url = reader.result;
      }
      file.id = file.size.toString() + new Date().getTime();
      if (onChange) {
        onChange(file);
      }
      if (callback) {
        callback(file);
      }
    };
  };

  const plusItem = <i className="plus-icon">+</i>;

  return (
    <div className="uploader">
      <div onClick={triggerInputUpload} className="upload-item">
        <span className="upload-button">
          <input
            type="file"
            ref={uploadInput}
            style={{ display: 'none' }}
            onChange={onFileChange}
            multiple={multiple}
            {...rest}
          />
          {renderPlusItem ? renderPlusItem() : plusItem}
        </span>
      </div>
    </div>
  );
};

export default Uploader;
