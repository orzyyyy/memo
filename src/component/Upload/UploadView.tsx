import React, { useEffect } from 'react';
import './css/UploadView.css';
import { isImageUrl } from './utils';
import Uploader, { UploaderProps, UploadFile } from './Uploader';
import classNames from 'classnames';

export interface UploadViewProps extends UploaderProps {
  fileList?: UploadFile[];
  onPress?: (file: UploadFile) => void;
  onClick?: (file: UploadFile) => void;
  uploaderVisible?: boolean;
}

const UploadView = ({
  fileList = [],
  onChange,
  accept,
  multiple,
  onClick,
  onPress,
  uploaderVisible,
}: UploadViewProps) => {
  let timer: ReturnType<typeof setTimeout>;

  useEffect(() => () => {
    clearTimer();
  });

  const handleViewTouchStart = (file: UploadFile) => {
    timer = setTimeout(() => {
      if (onPress) {
        onPress(file);
      }
    }, 800);
  };

  const clearTimer = () => {
    clearTimeout(timer);
  };

  return (
    <div className="uploadView">
      <div className="upload-view-list">
        {fileList.map(item => {
          const { url, id, name, error } = item;

          if (isImageUrl(url)) {
            return (
              <div
                className={classNames({ 'upload-item': true, 'upload-item-error': error })}
                key={`upload-item-${id}`}
                onTouchStart={() => handleViewTouchStart(item)}
                onTouchMove={clearTimer}
                onTouchEnd={clearTimer}
                onClick={() => onClick && onClick(item)}
              >
                <img src={url} />
              </div>
            );
          }

          return (
            <div className="upload-item" key={`upload-item-${id}`}>
              {name}
            </div>
          );
        })}
        {uploaderVisible && <Uploader fileList={fileList} onChange={onChange} accept={accept} multiple={multiple} />}
      </div>
    </div>
  );
};

export default UploadView;
