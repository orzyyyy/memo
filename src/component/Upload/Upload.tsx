import React from 'react';
import UploadView, { UploadViewProps } from './UploadView';
import classNames from 'classnames';
import './css/Upload.css';

export interface UploadProps extends UploadViewProps {
  style?: React.CSSProperties;
  className?: string;
}

const Upload = (props: UploadProps) => {
  const { style, className, ...rest } = props;
  return (
    <div className={classNames('upload', className)} style={style}>
      <UploadView {...rest} />
    </div>
  );
};

export default Upload;
