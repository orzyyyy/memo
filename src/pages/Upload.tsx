import React, { useState } from 'react';
import './css/upload.css';
import { UploadFile } from '../component/Upload/Uploader';
import Upload from '../component/Upload';
import Button from '../component/Button';
import { BaseDocumentProps } from './SlicingImage';

const UploadView = ({ innerHTML, className }: BaseDocumentProps) => {
  const [fileList, setFileList] = useState([] as UploadFile[]);
  const [response, setResponse] = useState('');

  const onChange = (changedFileList: UploadFile[]) => {
    setFileList([...fileList, ...changedFileList]);
  };

  const onSubmit = async () => {
    const formData: FormData = new FormData();
    for (const item of fileList) {
      formData.append('file', item as any);
      formData.append('filename', item.name);
    }
    const response = await fetch('/api/memo/upload', { body: formData, method: 'POST' });
    const result = await response.text();
    setResponse(result);
  };

  return (
    <div className={`upload ${className}`}>
      <Upload fileList={fileList} onChange={onChange} uploaderVisible={fileList.length === 0} multiple />
      <Button onClick={onSubmit}>上传</Button>
      <>{response}</>
      <div dangerouslySetInnerHTML={{ __html: innerHTML }} />
    </div>
  );
};

export default UploadView;
