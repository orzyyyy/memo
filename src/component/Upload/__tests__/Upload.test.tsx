import React from 'react';
import { mount } from 'enzyme';
import Upload from '../Upload';

const fileList = [
  { id: '1', url: 'test1', error: false, name: 'test1' },
  { id: '2', url: 'test2', error: false, name: 'test2' },
];

describe('Upload', () => {
  it('render correctly', () => {
    const wrapper: any = mount(<Upload fileList={fileList} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('onClick should work correctly', () => {
    const onClick = jest.fn();
    const wrapper: any = mount(<Upload fileList={fileList} onClick={onClick} />);
    wrapper
      .find('.upload-item')
      .at(0)
      .simulate('click');
    expect(onClick).toBeCalledWith({ id: '1', url: 'test1', error: false, name: 'test1' });
  });

  it('status error should work correctly', () => {
    fileList[0].error = true;
    const wrapper: any = mount(<Upload fileList={fileList} />);
    expect(wrapper.find('.upload-item-error').length).toBe(1);
  });

  it('fileList should work correctly', () => {
    const wrapper: any = mount(<Upload fileList={fileList} />);
    wrapper.setProps({ fileList: [fileList[0]] });
    expect(wrapper.find('.upload-item').length).toBe(2);
  });
});
