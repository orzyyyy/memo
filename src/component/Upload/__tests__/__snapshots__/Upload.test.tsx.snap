// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload render correctly 1`] = `
<Upload
  fileList={
    Array [
      Object {
        "error": false,
        "id": "1",
        "name": "test1",
        "url": "test1",
      },
      Object {
        "error": false,
        "id": "2",
        "name": "test2",
        "url": "test2",
      },
    ]
  }
>
  <div
    className="upload"
  >
    <UploadView
      fileList={
        Array [
          Object {
            "error": false,
            "id": "1",
            "name": "test1",
            "url": "test1",
          },
          Object {
            "error": false,
            "id": "2",
            "name": "test2",
            "url": "test2",
          },
        ]
      }
    >
      <div
        className="uploadView"
      >
        <div
          className="upload-view-list"
        >
          <div
            className="upload-item"
            key="upload-item-1"
            onClick={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
          >
            <img
              src="test1"
            />
          </div>
          <div
            className="upload-item"
            key="upload-item-2"
            onClick={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
          >
            <img
              src="test2"
            />
          </div>
          <Uploader
            fileList={
              Array [
                Object {
                  "error": false,
                  "id": "1",
                  "name": "test1",
                  "url": "test1",
                },
                Object {
                  "error": false,
                  "id": "2",
                  "name": "test2",
                  "url": "test2",
                },
              ]
            }
          >
            <div
              className="uploader"
            >
              <div
                className="upload-item"
                onClick={[Function]}
              >
                <span
                  className="upload-button"
                >
                  <input
                    onChange={[Function]}
                    style={
                      Object {
                        "display": "none",
                      }
                    }
                    type="file"
                  />
                  <i
                    className="plus-icon"
                  >
                    +
                  </i>
                </span>
              </div>
            </div>
          </Uploader>
        </div>
      </div>
    </UploadView>
  </div>
</Upload>
`;
